artichoke-backend/src/extn/core/string/ffi.rs

```diff
@@ -27,20 +27,7 @@ use crate::value::Value;
 #[no_mangle]
 unsafe extern "C" fn mrb_str_new_capa(mrb: *mut sys::mrb_state, capa: usize) -> sys::mrb_value {
     unwrap_interpreter!(mrb, to => guard);
-
-    let alloc_capacity = if let Some(capacity) = capa.checked_add(1) {
-        capacity
-    } else {
-        let err = ArgumentError::with_message("string capacity too large");
-        error::raise(guard, err);
-    };
-    // SAFETY: mruby assumes strings are allocated with `capacity = capa + 1`
-    // and that last byte is NUL.
-    let mut result = String::with_capacity(alloc_capacity);
-    let ptr = result.as_mut_ptr();
-    let last = ptr.add(capa);
-    *last = 0;
-
+    let result = String::with_capacity(capa);
```

Comment:
the behavior here is different than above and different than mruby.

mruby always sets the last byte of the allocation to NUL, whereas this code will set the first byte (the beginning of the spare capacity).

https://github.com/artichoke/artichoke/blob/16c869a9ad29acfe143bfcc011917ef442ccac54/artichoke-backend/vendor/mruby/src/string.c#L36-L38

I think probably the right fix is to set the first _and_ last bytes of the `Vec`'s spare capacity to zero.